### ISSUE TITLE
Training on a GPU and other issues

### ISSUE THREAD

**Issue by mrezasoltani** [Zhongyang-debug/Sandglasset-A-Light-Multi-Granularity-Self-Attentive-Network-For-Time-Domain-Speech-Separation#2]
Have you ever tried to train the model according to the hyperparameters of the paper on a single GPU? if so, what was the GPU and its memory you have used? I tried to train my own code and yours for the LibriMix dataset on a GPU T4 with 22GB memory, however, both your model and mine output "CUDA out of memory" error. The only setup I could make the model work is when you go with 2 blocks, and choose your hyperparameters very small, and the batch size equals 2. Apparently, in this setup, no success will be made for separating the LibriMix signals.

BTW: the encoding positional has some bugs.
         No module called adamp has been defined.
         in the train() file, in line 46, the positional argument of kernel-size has been missed.

**Comment by Zhongyang-debug** [comment]
I have tried to train the model, and the parameter setting of batch_size is small.

**Comment by Zhongyang-debug** [comment]
"No module called adamp has been defined."
pip3 install adamp
from adamp import AdamP
from adamp import SGDP

**Comment by Zhongyang-debug** [comment]
"in the train() file, in line 46, the positional argument of kernel-size has been missed."
This part has been updated.

**Comment by mrezasoltani** [comment]
> I have tried to train the model, and the parameter setting of batch_size is small.

For what dataset? How many GPUs did you use? What were your batch size and other hyper-parameters? 

As I mentioned, unless you go with a very small batch size (like batch size = 2), and select 2 Sandglasset blocks, and other things small. I can run it. Otherwise, I get "CUDA out of memory" with one GPU of NIVIDA Tesla 4 with 22 GB memory.

**Comment by Zhongyang-debug** [comment]
My dataset is WSJ0-2mix.batch_size=1，kernel_size=16，6 sandglasset blocks.

**Comment by mrezasoltani** [comment]
How many GPUs did you use?
The memory of the GPU you have used?

**Comment by Zhongyang-debug** [comment]
Only one GPU with 12GP memory.

**Comment by mrezasoltani** [comment]
> Only one GPU with 12GP memory.

Thanks

**Comment by mrezasoltani** [comment]
With batch size=1, did you see any improvement in training? Also, are you uing adam optimizer?

**Comment by Zhongyang-debug** [comment]
The advantage of this model is lightweight.I use adamp optimizer.

**Comment by mrezasoltani** [comment]
Even with batch size=1, do you see the error decreases, and the model can learn something?

**Comment by Zhongyang-debug** [comment]
For most models of speech separation, we can only use batch_siz=1, but it does have an effect.

**Comment by Zhongyang-debug** [comment]
I haven’t encountered this error before. Try to change the parameters. The two data sets should be similar.

**Comment by mrezasoltani** [comment]
No worries. Let you know once I have more details.

**Comment by Zhongyang-debug** [comment]
Thank you, it is very good to communicate with you.

**Comment by mrezasoltani** [comment]
Sure. me too.

**Comment by mrezasoltani** [comment]
This is my understanding from the paper and your code:
in_channels       # E         in the paper: 256
out_channels     # D        in the paper: 128
kernel_size       # M         in the paper: 4
length              # K          in the paper: 256
hidden_channels  # H      in the paper: 128
num_heads            # J     in the paper: 8
cycle_amount     # N       in the paper: 6

Is this correct? I suggest you to annotate your "class Sandglasset(nn.Module)" with these letters to help readers.

Thanks

**Comment by Zhongyang-debug** [comment]
Yes,thank you for your opinion, a larger kernel_size is better, otherwise it will easily exceed the memory.

**Comment by mrezasoltani** [comment]
Sure. Another issue in loading from a saved checkpoint:

First: the commented lines in _reset() need to be uncommented as the loaded model has been wrapped up by nn.DataParallel(model).

Second, even with this, if one tries to laod a model from a saved checkpoint (by setting config["save_load"]["continue_from"])   a "Cuda out of memory" error happens.

**Comment by Zhongyang-debug** [comment]
“First: the commented lines in _reset() need to be uncommented as the loaded model has been wrapped up by nn.DataParallel(model).”

This part of the code has been updated.

**Comment by Zhongyang-debug** [comment]
"Second, even with this, if one tries to laod a model from a saved checkpoint (by setting config["save_load"]["continue_from"]) a "Cuda out of memory" error happens."

That's the model itself.

You can try kernel_size = 16, batch_size = 1, 4 sandglasset blocks.
